daemon/cmd/devices.go

```diff
@@ -0,0 +1,296 @@
+	} else {
+		for _, dev := range option.Config.Devices {
+			dm.devices[dev] = true
+		}
+	}
+
+	detectDirectRoutingDev := option.Config.EnableNodePort
+	if option.Config.DirectRoutingDevice != "" {
+		dm.devices[option.Config.DirectRoutingDevice] = true
+		detectDirectRoutingDev = false
+	}
+
+	detectIPv6MCastDev := option.Config.EnableIPv6NDP
+	if option.Config.IPv6MCastDevice != "" {
+		dm.devices[option.Config.IPv6MCastDevice] = true
+		detectIPv6MCastDev = false
+	}
+
+	k8sNodeDev := ""
+	k8sNodeLink, err := findK8SNodeIPLink()
+	if err == nil {
+		k8sNodeDev = k8sNodeLink.Attrs().Name
+		dm.devices[k8sNodeDev] = true
+	}
+
+	if detectDirectRoutingDev {
+		// If only one device found, use that one. Otherwise use the device with k8s node IP.
+		if len(dm.devices) == 1 {
+			for dev := range dm.devices {
+				option.Config.DirectRoutingDevice = dev
+				break
+			}
+		} else if k8sNodeLink != nil {
+			option.Config.DirectRoutingDevice = k8sNodeDev
+		} else {
+			return fmt.Errorf("Unable to determine direct routing device. Use --%s to specify them",
+				option.DirectRoutingDevice)
+		}
+		log.Infof("Detected %s=%s", option.DirectRoutingDevice, option.Config.DirectRoutingDevice)
+	}
+
+	if detectIPv6MCastDev {
+		if k8sNodeLink != nil && k8sNodeLink.Attrs().Flags&net.FlagMulticast != 0 {
+			option.Config.IPv6MCastDevice = k8sNodeDev
+			log.Infof("Detected %s=%s", option.IPv6MCastDevice, option.Config.IPv6MCastDevice)
+		} else {
+			return fmt.Errorf("Unable to determine Multicast device. Use --%s to specify them",
+				option.IPv6MCastDevice)
+		}
+	}
+
+	// TODO(JM): Remove mutation of option.Config.Devices as it is unsafe. A follow up will add support
+	// for runtime detection of new devices and will switch cilium agent to use GetDevices() instead.
+	option.Config.Devices = dm.getDevices()
+	log.WithField(logfields.Devices, option.Config.Devices).Info("Detected devices")
+
+	return nil
+}
+
+// GetDevices returns the current list of devices Cilium should attach programs to.
+func (dm *DeviceManager) GetDevices() []string {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	return dm.getDevices()
+}
+
+func (dm *DeviceManager) getDevices() []string {
+	devs := make([]string, 0, len(dm.devices))
+	for dev := range dm.devices {
+		devs = append(devs, dev)
+	}
+	sort.Strings(devs)
+	return devs
+}
+
+// isViableDevice returns true if the given link is usable and Cilium should attach
+// programs to it.
+func (dm *DeviceManager) isViableDevice(l3DevOK bool, link netlink.Link) bool {
+	name := link.Attrs().Name
+
+	// Do not consider any of the excluded devices.
+	for _, p := range excludedDevicePrefixes {
+		if strings.HasPrefix(name, p) {
+			log.WithField(logfields.Device, name).Debugf("Skipping device as it has excluded prefix '%s'", p)
+			return false
+		}
+	}
+
+	// Skip slave devices.
+	if link.Attrs().RawFlags&unix.IFF_SLAVE != 0 {
+		log.WithField(logfields.Device, name).Debugf("Skipping slave device")
```

Comment:
I've not heard of "secondary device". We're testing whether the SLAVE flag is set, so we should talk about using the right terms? Or if we need to avoid that particular word we could use "lower" or "nested" device instead?

Reply:
This is now "bonded device"